### PR TITLE
Feat/ticks snipes

### PR DESCRIPTION
**Pull request by adhusson**
Snipes are now tick-based

New file: `MgvHelpers.sol` gives tools to call `snipesByVolume` or to manually convert a `targets` array to ticks. These utilities are now used by tests.

### CODE REVIEW

File: src/MgvHelpers.sol [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: Unlicense
+
+/* `MgvLib` contains data structures returned by external calls to Mangrove and the interfaces it uses for its own external calls. */
```
Comment (espendk) [inline src/MgvHelpers.sol]: This comment seems unrelated to this file - perhaps copy/paste leftover?

File: src/MgvOfferTaking.sol [hunk 2]
```diff
@@ -310,8 +310,9 @@ abstract contract MgvOfferTaking is MgvHasOffers {
   /* ## Snipes */
   //+clear+
 
-  /* `snipes` executes multiple offers. It takes a `uint[4][]` as penultimate argument, with each array element of the form `[offerId,takerWants,takerGives,offerGasreq]`. The return parameters are of the form `(successes,snipesGot,snipesGave,bounty,feePaid)`. 
+  /* `snipes` executes multiple offers. It takes a `uint[4][]` as penultimate argument, with each array element of the form `[offerId,tick,fillVolume,offerGasreq]`. The return parameters are of the form `(successes,snipesGot,snipesGave,bounty,feePaid)`.
```
Comment (espendk) [inline src/MgvOfferTaking.sol:313]: `tick` is an `int` in the other interfaces, so slightly surprising for it to a `uint` here. I think it's fine, as snipe will be a specialist call for cleaning, but we should probably call it out explicitly.

File: src/MgvOfferTaking.sol [hunk 3]
```diff
@@ -392,16 +393,16 @@ abstract contract MgvOfferTaking is MgvHasOffers {
...
-          mor.fillVolume = mor.fillWants ? wants : gives;
-          /* Wants gets normalized to the closest tick */
-          // note that `takerGives=0` is not possible here
-          // Tick tt = TickLib.tickFromVolumes(gives,wants);
-          // sor.tick = TickLib.tickFromVolumes(gives,wants);
-          mor.maxTick = TickLib.tickFromTakerVolumes(gives, wants);
+          {
+            Tick tick = Tick.wrap(int(targets[i][1]));
+            require(TickLib.inRange(tick), "mgv/snipes/tick/outOfRange");
+            mor.maxTick = tick;
```
Comment (espendk) [inline src/MgvOfferTaking.sol:399]: I was expecting snipe to only work if the offer was at the specified tick. But it seems we also execute offers where the price is better than expected?

File: src/periphery/MgvCleaner.sol [hunk 4]
```diff
@@ -30,7 +31,8 @@ contract MgvCleaner {
     returns (uint bal)
   {
     unchecked {
-      (uint successes,,,,) = MGV.snipesFor(outbound_tkn, inbound_tkn, targets, fillWants, msg.sender);
+      (uint successes,,,,) =
+        MgvHelpers.snipesForByVolume(address(MGV), outbound_tkn, inbound_tkn, targets, fillWants, msg.sender);
```
Comment (espendk) [inline src/periphery/MgvCleaner.sol:35]: Assume the MgvCleaner API was left unchanged in anticipation of it being removed?

**Review by espendk — APPROVED** [review]
LGTM. Added a few comments.

I think most of the snipe tests are more natural when using the tick API, so I've taken the liberty of migrating them. With the exception of the MgvCleaner tests (and API) which will soon be irrelevant anyway.

As discussed, sniping will become a specialized API for cleaning. So I think we should consider removing the utility methods from MgvHelpers. But perhaps we could move the other ByVolume versions of the maker and taker APIs there?

**Comment by espendk** [comment]
I forgot to mention one thing: Creating the `targets` for `snipes` has become somewhat clunky due to ticks:

```solidity
wrap_dynamic([ofr, uint(Tick.unwrap(tick)), 1 ether, 300_000])
```

I couldn't come up with something better outside of creating a utility method for it. Do you see better options?

File: src/MgvHelpers.sol (hunk 1 above)
Comment (adhusson) [reply src/MgvHelpers.sol]: yes

File: src/MgvOfferTaking.sol (hunk 3 above)
Comment (adhusson) [reply src/MgvOfferTaking.sol:399]: yes, the idea is that the taker specifies the offer tick as a protection against the offer having changed. But if the offer has _improved_ its price the taker needs no protection

**Comment by adhusson** [comment]
> I forgot to mention one thing: Creating the `targets` for `snipes` has become somewhat clunky due to ticks:
> 
> ```solidity
> wrap_dynamic([ofr, uint(Tick.unwrap(tick)), 1 ether, 300_000])
> ```
> 
> I couldn't come up with something better outside of creating a utility method for it. Do you see better options?

Not with arrays, but we may figure something out with structs!

File: src/MgvOfferTaking.sol (hunk 2 above)
Comment (adhusson) [reply src/MgvOfferTaking.sol:313]: will add a comment

File: src/periphery/MgvCleaner.sol (hunk 4 above)
Comment (adhusson) [reply src/periphery/MgvCleaner.sol:35]: just minimizing changes :p
